### DISCUSSION TITLE
Using the `SourceMapper` fails with `Should be able to convert to usize: TryFromIntError(())`

### DISCUSSION THREAD

**Discussion by stahlbauer** [swc-project/swc#6206] [Q&A]
I now have a SWC plugin ready for testing 🎉 and are facing some problems 💥.

**Question(s) 1**: What are general best practices for debugging SWC plugins? For example, how to get to meaningful stack traces? Which debugger to use? Is it possible to debug into the code from a NodeJS environment?

**Question 2**: I am facing a concrete failure in `swc_plugin_runnter-0.77.4/src/memory_interop.rs:37.10` when running the plugin from a NodeJs environment. Any ideas?: 

```
thread '<unnamed>' panicked at 'Should be able to convert to usize: TryFromIntError(())', /usr/local/cargo/registry/src/github.com-1ecc6299db9ec823/swc_plugin_runner-0.77.4/src/memory_interop.rs:37:10
stack backtrace:
   0:     0x7f92af26a7a0 - <unknown>
   1:     0x7f92af292e9e - <unknown>
   2:     0x7f92af265de5 - <unknown>
...
  80:     0x7f91f3312827 - <unknown>
Error: failed to handle: Should be able to convert to usize: TryFromIntError(())

Stack backtrace:
   0: <unknown>
   1: <unknown>
   2: <unknown>
   3: <unknown>
   4: <unknown>
   5: _ZN6v8impl12_GLOBAL__N_123FunctionCallbackWrapper6InvokeERKN2v820FunctionCallbackInfoINS2_5ValueEEE
   6: _ZN2v88internal12_GLOBAL__N_119HandleApiCallHelperILb0EEENS0_11MaybeHandleINS0_6ObjectEEEPNS0_7IsolateENS0_6HandleINS0_10HeapObjectEEESA_NS8_INS0_20FunctionTemplateInfoEEENS8_IS4_EENS0_16BuiltinArgumentsE
   7: _ZN2v88internal21Builtin_HandleApiCallEiPmPNS0_7IsolateE
   8: Builtins_CEntry_Return1_DontSaveFPRegs_ArgvOnStack_BuiltinExit
```

**Comment by kdy1** [comment]
Did you use 

```
strip = "symbols"
```
?

This is very bad for debugging.

I prefer unit testing & debugging via `test_fixture`

**Reply by stahlbauer** [reply]
I did not make use of `strip = "symbols"` and are working with a debug build---the `--release` flag is omitted.

**Reply by kdy1** [reply]
About the second question, what's the version of `swc_core` of your plugin.
Also, are you using `SourceMap` api?

**Reply by stahlbauer** [reply]
```
swc_core = {version = "0.40.1", features = [
  "plugin_transform",
  "common",
  "ecma_ast",
  "ecma_utils",
  "ecma_visit",
  "ecma_parser",
  "ecma_transforms",
  "ecma_codegen",
]}
```
I am using source maps in various ways. Did you have any specific crate in mind?

**Reply by kdy1** [reply]
Can you try adding `swc_common` with `concurrent`?

Just like https://github.com/swc-project/plugins/blob/d5499715a0d95218164ff5dccd63d65f2f5f2433/packages/emotion/Cargo.toml#L15

**Reply by stahlbauer** [reply]
Thanks. I tried it but was not able to observe any differences.

**Reply by kdy1** [reply]
What's your version of `@swc/core`?

**Reply by stahlbauer** [reply]
In my `package.json` I use `"@swc/core": "1.3"`. Which is currently locked to `1.3.4`.

**Reply by stahlbauer** [reply]
Also with the recent `1.3.9`, the error is raised.

**Reply by kdy1** [reply]
Did you try deleting `.swc` folder? It's used by swc to cache compiled wasm files

**Reply by stahlbauer** [reply]
Running the plugin from the test fixture works as desired. (With approximately the same inputs.)

**Reply by stahlbauer** [reply]
Deleting the `.swc` folder did not have an observable effect. (Along with rebuilding and re-running).

**Reply by kdy1** [reply]
cc @kwonoj Any ideas?

**Reply by kwonoj** [reply]
I saw this once when rkyv version / schema mismatch and underlying value becomes bogus. I'd check the actual value to see if that's the case.

**Reply by stahlbauer** [reply]
The version of `rkyv` used in the plugin and SWC is the same. Any pointers on how to start a debug session from VSCode along with a plugin/SWC configuration?

**Reply by kwonoj** [reply]
unfortunately there's no way to attach debuggers into plugin's stack currently.

**Reply by stahlbauer** [reply]
> I'd check the actual value to see if that's the case.

I thought about debugging this (dumping the value making problems in `swc_plugin_runner-0.77.4/src/memory_interop.rs:37:10`). (If this brings any insights at all.)

**Reply by kwonoj** [reply]
You can't attach but still able to print values. Tried println?

**Reply by stahlbauer** [reply]
Would it make sense to run `swc_plugin_runner` with the plugin? (While `cargo test` in `swc_plugin_runner` currently does not execute any test.) Or is too much of the environment missing then? What is the role of `#[cfg(feature = "__rkyv")]`? Where will the feature be enabled?

**Reply by stahlbauer** [reply]
`println!` would also be fine. Then I have to figure out how to run a dev version of SWC with a plugin :-D

**Reply by kwonoj** [reply]
I'm not fully able to understand your question. If you are writing plugin you don't have direct way to deal with plugin runner from beginning as swc/core node.js binding should have it. If you are trying to run test inside of swc's repo itself for the plugin, that's not for the public usage.

**Reply by kwonoj** [reply]
And yes, if you want to print host side value you may need to build swc by yourself.

**Reply by kwonoj** [reply]
Overall, current debugging experience needs improvement regardless.

**Comment by stahlbauer** [comment]
In case you would have more pointers on fixing/debugging `Should be able to convert to usize: TryFromIntError(())` I would be thankful. 

I am also wondering how a minimal setup to reproduce the issue could look like. Maybe without the NodeJs part to reduce the complexity. Wouldn't the `swc_plugin_runner` be a good starting point?

**Reply by kdy1** [reply]
You can do the same thing we do for testing plugins.
Clone swc, and write a test in https://github.com/swc-project/swc/tree/caaf491da51b32424baf280513cf554c59aae72a/crates/swc_plugin_runner/tests and invoke it. It'll provide almost all information possible I guess

**Reply by stahlbauer** [reply]
For now, it is rather hard for me to set this up. Already running the test suite of SWC locally (as described in `CONTRIBUTING.md`) fails with test failures.

**Comment by stahlbauer** [comment]
By removing some of the transformation steps in my plugin I was able to get rid of the `Error: failed to handle: Should be able to convert to usize: TryFromIntError(())`. That is, the plugin seemed to have doing transformations that cause the trouble. Now, I will narrow down the manipulation a bit to identify the root cause.

**Reply by stahlbauer** [reply]
Calling `mapper.span_to_snippet`, where `mapper: rc<dyn SourceMapper + 'a>` from `metadata.source_map`, with `metadata:  TransformPluginProgramMetadata`, causes the problem somehow. Any ideas? @kwonoj

**Reply by stahlbauer** [reply]
Some of the calls to `span_to_snippet` succeed, and eventually, a call fails.

**Reply by kdy1** [reply]
You can check for `DUMMY_SP`

**Reply by stahlbauer** [reply]
Checking this with `span.is_dummy()` does not prevent the failure.

**Reply by stahlbauer** [reply]
What is the relationship of [plugin_source_map_proxy.rs](https://github.com/swc-project/swc/blob/main/crates/swc_plugin_proxy/src/source_map/plugin_source_map_proxy.rs) and [memory_interop.rs](https://github.com/swc-project/swc/blob/main/crates/swc_plugin_runner/src/memory_interop.rs) here? I can see many `wasm32` annotations in the former. How can a particular call make the plugin fail?

**Reply by stahlbauer** [reply]
What about concurrency?

**Comment by stahlbauer** [comment]
The following seems to be sufficient to reproduce the problem:
```
#[plugin_transform]
pub fn process_transform(
    mut program: Program,
    metadata: TransformPluginProgramMetadata,
) -> Program {

    for i in 1..100000 {
        metadata.source_map.span_to_snippet(Span::new(BytePos(i), BytePos(i+10), SyntaxContext::empty()));
    }
```

**Reply by stahlbauer** [reply]
For now, I have created an issue: https://github.com/swc-project/swc/issues/6249
